tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_controller_component.ts

```diff
@@ -194,4 +194,31 @@ export class LinkedTimeFobControllerComponent {
     // No fobs can pass the lowest step in this graph.
     return 0;
   }
+
+  stepTyped(fob: Fob, step: number) {
+    let newLinkedTime = this.linkedTime;
+    if (!newLinkedTime.end) {
+      newLinkedTime.start.step = step;
+      this.onSelectTimeChanged.emit(newLinkedTime);
+      return;
+    }
+
+    if (fob === Fob.START) {
+      newLinkedTime.start.step = step;
+      if (step > newLinkedTime.end.step) {
+        newLinkedTime.start.step = newLinkedTime.end.step;
+        newLinkedTime.end.step = step;
+      }
+    }
+
+    if (fob === Fob.END) {
+      newLinkedTime.end.step = step;
+      if (step < newLinkedTime.start.step) {
+        newLinkedTime.end.step = newLinkedTime.start.step;
+        newLinkedTime.start.step = step;
+      }
+    }
+
+    this.onSelectTimeChanged.emit(newLinkedTime);
```

Comment:
FWIW, to sort of put all my other comments together, I think this entire function could be rewritten as:

```

  stepTyped(fob: Fob, step: number) {
    let newLinkedTime = {...this.linkedTime};
    if (fob === Fob.START) {
      newLinkedTime.start = {step};
    } else if (fob === Fob.END) {
      newLinkedTime.end = {step};
    }

    if (
      newLinkedTime.end !== null &&
      newLinkedTime.start.step > newLinkedTime.end.step
    ) {
      // Start is now greater than end - flip them.
      newLinkedTime = {
        start: newLinkedTime.end,
        end: newLinkedTime.start,
      };
    }

    this.onSelectTimeChanged.emit(newLinkedTime);
  }
```

Some advantages are:
No need to early return.
Only single place where you might call emit().
Only single place where values are checked if they need to be swapped and are swapped.

Reply:
I have always liked early returns. However, this is definitely much cleaner logic. Thanks!